### ISSUE TITLE
how to implement ReserveNowRequest with OCPP v1.6J

### ISSUE THREAD

**Issue by bdevia** [ChargeTimeEU/Java-OCA-OCPP#309]
hi, I need help to correctly implement the ReserveNowRequest method. I have been doing several tests but when sending the request it throws an unsupported exception, so I think I am making the request wrong.

I attach the code:
```
public class ReservationRequest extends RequestHandler {
    
    public CompletableFuture<ResponseEntity<ObjectNode>> reserveNow(JsonNode params){
        if(params.has(OcppEnum.CHARGE_POINT_ID.getName()) && params.has(OcppEnum.CONNECTOR_ID.getName()) && params.has(OcppEnum.ID_TAG.getName()) && params.has(OcppEnum.RESERVATION_ID.getName())){
            String chargePointId = params.get(OcppEnum.CHARGE_POINT_ID.getName()).asText();
            Integer connectorId = params.get(OcppEnum.CONNECTOR_ID.getName()).asInt();
            String idTag = params.get(OcppEnum.ID_TAG.getName()).asText();
            Integer reservationId = params.get(OcppEnum.RESERVATION_ID.getName()).asInt();

            ZonedDateTime nowTime = ZonedDateTime.now();      // Por defecto la reservacion dura 5 minutos;
            ZonedDateTime expiryDate = nowTime.plus(5, ChronoUnit.MINUTES);

            System.out.println(chargePointId + " " + connectorId + " " + idTag + " " + expiryDate + " " + reservationId );

            ReserveNowRequest reservationNowRequest = new ReserveNowRequest(connectorId, expiryDate, idTag, reservationId);
            printSendingRequest(chargePointId);

            return sendRequest(chargePointId, reservationNowRequest);
        }
        else{
            return badRequest("'" + OcppEnum.CHARGE_POINT_ID.getName() + "', '" + OcppEnum.CONNECTOR_ID.getName() + "', '" + OcppEnum.ID_TAG.getName() + "' and '" + OcppEnum.RESERVATION_ID.getName() +  "' are mandatory parameters");
        }
    }
}
```
sendRequest is implemented here:
```
public class RequestHandler extends ErrorHandler{
    
    private final WebSocketServer websocket = WebSocketServer.getInstancia();

    protected CompletableFuture<ResponseEntity<ObjectNode>> sendRequest(String chargePointId, Request request){
        
        /* VALIDACION DE CHARGE POINT ID */
        if(!websocket.getEventHandler().getSessionIds().containsKey(chargePointId)){
            return errorChargePointId(chargePointId);
        }
        else{
            UUID sessionIndex = websocket.getEventHandler().getSessionIds().get(chargePointId);
            CompletableFuture<ResponseEntity<ObjectNode>> resultFuture = new CompletableFuture<>();
            long timeoutDuration = 5000;
            try {
                CompletableFuture.runAsync(() -> {
                    try {
                        Thread.sleep(timeoutDuration);
                        resultFuture.complete(errorHandler(new Exception("Request Timeout"), HttpStatus.REQUEST_TIMEOUT));
                    } catch (Throwable t) {
                        resultFuture.complete(errorHandler(t, HttpStatus.INTERNAL_SERVER_ERROR));
                    }
                });

                websocket.getServer().send(sessionIndex, request).whenComplete((confirmation, t) -> {
                    if(t != null){
                        t.printStackTrace();
                        resultFuture.complete(errorHandler(t, HttpStatus.INTERNAL_SERVER_ERROR));
                    }
                    else{
                        System.out.println("Result: " + confirmation);
                        System.out.println();

                        ObjectNode response = new ObjectMapper().createObjectNode();
                        response.set("response", new ObjectMapper().valueToTree(confirmation));
                        resultFuture.complete(new ResponseEntity<>(response, HttpStatus.OK));
                    }
                });
            } catch (Throwable t) {
                t.printStackTrace();
                resultFuture.complete(errorHandler(t, HttpStatus.INTERNAL_SERVER_ERROR));
            }
            return resultFuture;
        }
    }
```

I have implemented other Request methods in the same way and they work correctly, however this particular method is not working for me.
This is the error it gives me:
```
feb. 09, 2024 5:26:57 P.áM. org.apache.catalina.core.ApplicationContext log
INFO: Initializing Spring DispatcherServlet 'dispatcherServlet'
JN246801357924 1 testing 2024-02-09T17:31:57.543517900-03:00[America/Santiago] 8
Sending a reserveNow resquest for chargePoint JN246801357924
eu.chargetime.ocpp.UnsupportedFeatureException
        at eu.chargetime.ocpp.Server.send(Server.java:200)
        at eu.chargetime.ocpp.JSONServer.send(JSONServer.java:164)
        at com.example.ws.ocpp16.handler.RequestHandler.sendRequest(RequestHandler.java:41)
        at com.example.ws.ocpp16.handler.request.ReservationRequest.reserveNow(ReservationRequest.java:34)
        at com.example.controller.ApiController.reserveNow(ApiController.java:74)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:261)
        at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:189)
        at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:118)
        at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:917)
        at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:829)
        at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:87)
        at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1089)
        at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:979)
        at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1014)
        at org.springframework.web.servlet.FrameworkServlet.doPost(FrameworkServlet.java:914)
        at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:590)
        at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:885)
        at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:658)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:205)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:149)
        at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:51)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:174)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:149)
        at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:100)
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:174)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:149)
        at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:93)
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:174)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:149)
        at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201)
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:174)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:149)
        at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:167)
        at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:90)
        at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:482)
        at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:115)
        at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:93)
        at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:74)
        at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:340)
        at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:391)
        at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:63)
        at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:896)
        at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1744)
        at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:52)
        at org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1191)
        at org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:659)
        at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
        at java.base/java.lang.Thread.run(Thread.java:842)
```
 has anyone encountered this problem?

**Comment by bdevia** [comment]
Hi,

I have already solved the problem! As mentioned in the thread: https://github.com/ChargeTimeEU/Java-OCA-OCPP/issues/103, profiles must be manually added to the server with the "addFeatureProfile" method. This way I resolved the exception by setting the following line in my server's constructor:
```
this.server.addFeatureProfile(new ServerReservationProfile());
```

I hope it helps someone!

**Comment by TVolden** [comment]
Hi @bdevia 

Sorry I didn't manage to reply to your issue, I'm swamped with work these days.
It's nice that you found a solution, and thanks for sharing.

**Comment by brandongallagher** [comment]
In case it is helpful to someone else, this is how i added the TriggerMessageRequest feature in my JsonServer.  This is based on the JsonServer in the 1.6J example server:
`@Configuration
@Slf4j
public class JsonServerConfig {

    @Bean
    public JSONServer jsonServer(ServerCoreProfile core, ServerRemoteTriggerProfile remoteTriggerProfile) {
        // create the server with the core features
        JSONServer jsonServer = new JSONServer(core);

        // add the remote trigger profile so we can use this feature's functionality
        jsonServer.addFeatureProfile(remoteTriggerProfile);

        return jsonServer;
    }

}`

**Comment by rpseng** [comment]
Hello, I had to add something similar when implementing smart charging in a test server.

My question is, why is this needed in the first place? AFAIK only clients should announce the supported feature profiles, not servers. I failed to find in the spec a section mentioning that servers should disclose the features they implement. To me, it looks unnecessary.

Despite this minor concern, I want to emphasize that this library is immensely valuable as the basis for other projects. Keep up the great work.
